### PR TITLE
Add mypy to CI

### PR DESCRIPTION
**Pull request by kbressem**
# Summary
1. This PR adds `mypy` to the CI as a static type checker and fixes files to resolve new type errors. 

2. `get_loss` was updated to accept not only loss functions from `monai.losses` but also from `torch.nn`. 

## Changes
- `ci.yml`: adds `mypy` after flake8. Removes python 3.7 because of an apparent [error in `shutil.copytree`](https://github.com/kbressem/trainlib/actions/runs/3312627546/jobs/5469535287#step:6:29) 
- `mypy` options were added to the `setup.cfg`
- Type errors were corrected throughout the library
- `trainlib.loss.get_loss` was adapted to also accept losses from pytorch.
- tests for loss, optimizer and model were moved to separate files. Tests for loss were enhanced.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,7 +10,8 @@ pretty:
 	black --line-length 120 .
```
Comment (markus-hinsche) [inline Makefile:10]: Consider running pretty in CI

https://github.com/rAIdiance/chest/blob/main/.github/workflows/ci.yml#L38

File: tests/test_utils.py [hunk 2]
```diff
@@ -20,3 +21,9 @@ class TestCase(unittest.TestCase):
         [torch.zeros(1, 3, 3, 3), torch.ones(1, 3, 3, 3), torch.full((1, 3, 3, 3), 2)],
         1,
     )
+
+
+def func_req_additional_args(func):
+    "Checks if non-default arguments exist in function"
```
Comment (markus-hinsche) [inline tests/test_utils.py:27]: """docstring"""

more occurances in this file

File: tests/test_utils.py [hunk 3]
```diff
@@ -20,3 +21,9 @@ class TestCase(unittest.TestCase):
         [torch.zeros(1, 3, 3, 3), torch.ones(1, 3, 3, 3), torch.full((1, 3, 3, 3), 2)],
         1,
     )
```
Comment (markus-hinsche) [inline tests/test_utils.py:23]: TestCase -> TestCaseConfig

File: tests/test_utils.py [hunk 4]
```diff
@@ -20,3 +21,9 @@ class TestCase(unittest.TestCase):
...
+
```
Comment (markus-hinsche) [inline tests/test_utils.py:24]: config as class variable might not support multiple different configs. 

Maybe using instance variables would be safer

File: tests/test_loss.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+LOSSES = monai_losses + torch_losses
+
+
+class TestLoss(TestCase):
+    @parameterized.expand(LOSSES)
+    def test_init(self, loss_name):
+        "Simple test if losses can be initialized"
+        config = self.config
+        config.loss = {loss_name: {}}
```
Comment (markus-hinsche) [inline tests/test_loss.py:24]: Does this change a class variable?

File: tests/test_loss.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+monai_losses = [x for x in monai_losses if not func_req_additional_args(getattr(monai.losses, x))]
+torch_losses = [x for x in torch_losses if not func_req_additional_args(getattr(torch.nn, x))]
+
+LOSSES = monai_losses + torch_losses
+
+
+class TestLoss(TestCase):
+    @parameterized.expand(LOSSES)
+    def test_init(self, loss_name):
+        "Simple test if losses can be initialized"
```
Comment (markus-hinsche) [inline tests/test_loss.py:22]: """

File: tests/test_loss.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        config = self.config
```
Comment (markus-hinsche) [inline tests/test_loss.py:23]: Does this copy or point to the reference?

File: tests/test_utils.py [hunk 8]
```diff
@@ -20,3 +21,9 @@ class TestCase(unittest.TestCase):
...
+
+def func_req_additional_args(func):
```
Comment (markus-hinsche) [inline tests/test_utils.py:26]: This function name could sound more boolean. 

e.g. `does_func_have_additional_args()`

File: trainlib/loss.py [hunk 9]
```diff
@@ -17,6 +21,12 @@ def get_loss(config: dict):
     """
     loss_type = list(config.loss.keys())[0]
     loss_config = config.loss[loss_type]
-    loss_fun = getattr(monai.losses, loss_type)
+    if hasattr(monai_losses, loss_type):
+        loss_fun = getattr(monai_losses, loss_type)
+    elif hasattr(torch.nn, loss_type):
+        loss_fun = getattr(torch.nn, loss_type)
+    else:
+        raise AttributeError(f"Lossfunction {loss_type} is not in `monai.losses` or `torch.nn`")
```
Comment (markus-hinsche) [inline trainlib/loss.py:29]:
```suggestion
        raise AttributeError(f"Loss function '{loss_type}' is neither in `monai.losses` nor in `torch.nn`")
```

**Review by markus-hinsche — APPROVED** [review]
LGTM! Thanks!

I added a few comments for improvements

For the future you can consider: split into 2 PRs (loss and mypy are non-overlapping topics)

**Comment by markus-hinsche** [comment]
https://stackoverflow.com/questions/63661976/problems-using-shutil-copytree-in-python#comment116491643_63662069

If you wanted to keep py 3.7.:
```
if dirs_exists:
    <your code without dirs_exist_ok>
```

File: tests/test_loss.py (hunk 5 above)
Comment (kbressem) [reply tests/test_loss.py:24]: The default config.loss is: 
```python
{'DiceCELoss': {'include_background': False, 'softmax': True, 'to_onehot_y': True}}
```
For testing, it is changed to the respective loss to test. E.g, if `DiceLoss` is tested, config.loss would be changed to: 

```python
{'DiceLoss': {}}
```
As I only pass loss functions that have default args during init, the empty dictionary is fine.

File: tests/test_loss.py (hunk 7 above)
Comment (kbressem) [reply tests/test_loss.py:23]: Not sure. It should copy... but probably points. Seems to work well regardless, as the same key is overwritten over and over again. But I'll explicitly copy the dict.

File: tests/test_utils.py (hunk 3 above)
Comment (kbressem) [reply tests/test_utils.py]: `TestCase` was removed and config, image etc. are now imported directly from `test_utils`

File: Makefile (hunk 1 above)
Comment (kbressem) [reply Makefile:10]: Black sorts imports slightly differently than isort, which would lead to problems in the CI. Do we use isort in rAIdiance?
